src/compiler/checker.ts

```diff
@@ -24618,15 +24617,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     // getApparentType can return _any_ type, since an indexed access or conditional may simplify to any other type.
                     // If that occurs and it doesn't simplify to an object or intersection, we'll need to restart `inferFromTypes`
                     // with the simplified source.
-                    if (apparentSource !== source && allowComplexConstraintInference && !(apparentSource.flags & (TypeFlags.Object | TypeFlags.Intersection))) {
-                        // TODO: The `allowComplexConstraintInference` flag is a hack! This forbids inference from complex constraints within constraints!
-                        // This isn't required algorithmically, but rather is used to lower the memory burden caused by performing inference
-                        // that is _too good_ in projects with complicated constraints (eg, fp-ts). In such cases, if we did not limit ourselves
-                        // here, we might produce more valid inferences for types, causing us to do more checks and perform more instantiations
-                        // (in addition to the extra stack depth here) which, in turn, can push the already close process over its limit.
-                        // TL;DR: If we ever become generally more memory efficient (or our resource budget ever increases), we should just
-                        // remove this `allowComplexConstraintInference` flag.
-                        allowComplexConstraintInference = false;

```

Comment:
quite wild that it was only ever set to `false` and never reset back to `true` 😅

Reply:
It's local to this infer stack, so it's not like the worst thing in the world.

Reply:
Aaah, makes sense - I missed that this is a closure variable.

Reply:
Yeah, it was basically used so we'd only explore one level of this kind of inference per `infer` call. So long as RWC no longer OOMs, we're good. We didn't need this to stop runaway memory utilization, but rather just to reduce it by like 5% because RWC at the time was *just* above memory limits (and we'd has issues with, eg, UMDWeb OOM'ing consistently) with the added inference, and looking into across-the-board memory utilization, rather than just limiting the most recent change, was painful.